src/walker.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Pipeline;
+  stepFn?: (step: Step) => Step;
+  commandFn?: (command: Command) => Command;
+}
+
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  const pipeline = new Pipeline(p.name);
+  pipeline.steps = [];
+  for (const step of p.steps) {
+    let newStep: PotentialStep | null = cloneDeep(step);
```

Comment:
this cloning assumes that the step is nothing more than a simple JSON object. I am not 100% sure this is currently true but definitely can't be guaranteed. If we had a generic `Clonable` interface that added a `clone(): this` method that has to be implemented, we could use that and make sure that the implementation itself ensures that a copy is correct.

Reply:
Good point, now I feel bad for suggesting the cloning, it's quite a lot of extra code for hiding the state as it is being mutated.

Are immutable objects a common pattern in javascript? Like https://immutable-js.com/. They're very common in Java, but if doing something like that is not going to be idiomatic, and callers might expect `pipeline.walk` to modify the pipeline in place, it might be worth making the interface more idiomatic and the implementation simpler.